### ISSUE TITLE
Different behavior between repo.remote().url and repo.remote().urls when dealing with UNC paths

### ISSUE THREAD

**Issue by HudsonMC16** [gitpython-developers/GitPython#1019]
Not sure if this is specifically realted to UNC paths or not, but here is what I'm observing. 

Clone a repo from a networked server to your local machine. Then:

```
>>> from Pathlib import Path
>>> path_str = local_repo.remote('origin').url
>>> print(path_str)
'\\\\\\\\network\\\\path\\\\to\\\\remote\\\\repo'
>>> Path(path_str).exists()
False
>>> Path(path_str).resolve()
WindowsPath('C:/network/path/to/remote/repo')
```
whereas:
```
>>> from pathlib import Path
>>> path_str = next(local_repo.remote('origin').urls)
>>> print(path_str)
'\\\\network\\path\\to\\remote\\repo'
>>> Path(path_str).exists()
True
>>> Path(path_str).resolve()
WindowsPath('//network/path/to/remote/repo')
```
Thanks!

**Comment by Byron** [comment]
Thanks a lot for posting this here!

I had a quick look and believe the reason is the following: When [asking for `urls`](https://github.com/gitpython-developers/GitPython/blob/24cd6dafc0008f155271f9462ae6ba6f0c0127a4/git/remote.py#L528-L534) git will be invoked to ask for its view of the world.

When [asking for `url`](https://github.com/gitpython-developers/GitPython/blob/24cd6dafc0008f155271f9462ae6ba6f0c0127a4/git/objects/submodule/base.py#L1119-L1121) it will use the potentially manipulated path that was used to instantiate the `Repo` instance.

The issue seems to be stemming from GitPython incorrectly transforming the UNC path, potentially treating it like some sort of Linux path.

**Comment by HudsonMC16** [comment]
So seems it would make sense for each function to access the url in the same way. Do we have a preference? Should the `.url` code be fixed to quit mangling the paths, or be rewritten to invoke git, similar to `.urls`?

**Comment by Byron** [comment]
I think it does make sense that their output is consistent, with `urls` being correct in this case.

Since calling implementing `url` like `next(urls)` is not the same both in terms of performance nor (probably) the result itself, I would clearly prefer `url` to be fixed to handle UNC paths correctly on windows. This might have the positive side-effect of fixing additional bugs that that stem from incorrect path handling.

**Comment by Byron** [comment]
Sorry for the `url` link, indeed it should rather have been a link to [this location](https://github.com/gitpython-developers/GitPython/blob/24cd6dafc0008f155271f9462ae6ba6f0c0127a4/git/remote.py#L430:L430), which is the fallback in case the attribute does not exist. In that case, it seems the URL is read directly from the configuration, `_config_reader.get(attr)`. That's interesting, as I would be surprised if GitPython would do any processing there. Can it be that the URL attribute is indeed an incorrect path?

**Comment by Byron** [comment]
Thanks for the research! This means that if we would use `"rt"` instead, the issue would somewhat magically disappear.

The reason the file is read as binary is that otherwise, the operating system (or Python, maybe) makes assumptions about the encoding of the text file, even though git itself doesn't know or care. Therefore reads can fail even though otherwise they won't.

GitPython obviously tries to decode the file as UTF-8 effectively on a line-by-line basis, which may fail as well so it isn't necessarily better.

Something I don't understand is how "RT" can clean up a path like "\\\\path\\dir" into "\\path\dir", which seems way more than a decoder would ever do.

In a way, I don't even understand why such a path would be written in the first place, unless backslash is a reserved character in git config files.

**Comment by HudsonMC16** [comment]
So, escape characters on different platforms and encodings make my head spin, but I expect this is a windows-only problem due to windows use of backslashes in paths. 

Even on a local repo, with no UNC complications, it looks like (under a utf-8 lens) git adds extra escape backslashes. So the entry in the config file looks like this in binary:
`\turl = C:\\\\path\\\\to\\\\repo\n`
and this in utf-8:
`\turl = C:\\path\\to\\repo\n`

So, I'm not sure that the gitpython configparser is decoding as utf-8. I'm still trying to follow the config lines through, but that's my best guess at this point.

**Comment by HudsonMC16** [comment]
I forgot to include, I think the reason the `'rt'` mode works, is because it automatically correctly escapes the backslashes, similar to decoding using the python `'unicode-escape'` encoding. So, for example: `b'\\\\'.decode('unicode-escape')` evaluates to `'\\'`

**Comment by Byron** [comment]
Awesome! `rt` is implemented by Python, not by the OS, it seems. And unicode-escaping needs backslashes to insert codepoints like `\uffff`, among others. All that seems to be done so that the configuration file itself does not actually have to be UTF-8 encoded, but works with ascii instead.

What would that mean for a fix? Doesn't that mean that one could simply [decode with 'unicode-escape'](https://github.com/gitpython-developers/GitPython/blob/ff4f970fa426606dc88d93a4c76a5506ba269258/git/config.py#L374) instead?

_I couldn't really find details on what's expected of git configuration files, but would have wanted to post this here to verify the solution above would be acceptable._

**Comment by HudsonMC16** [comment]
Yeah, I know what you mean about the git configuration files. I went looking for something also, but couldn't find anything. 

We could potentially decode with `'unicode-escape'` like you suggest. I guess we could also catch the 8 backslash pattern and replace it explicitly. That seems a little hacky, but less likely to accidentally break other stuff, I think. Another sort of hacky, but maybe more complete solution is to explicitly replace every four sequential binary backslashes with a single forward slash. Builtins and stdlib on windows (`pathlib.Path`, `open`, etc.) appear to work with strings structured this way.

**Comment by Byron** [comment]
Sometimes one has to dig deeper :D, so I found [this portion of the git source code](https://github.com/git/git/blob/master/config.c#L594:L594) which seems to mean that it's simply decoding these with `\\ ➡ \` .

Thus for us it might indeed be safest to do an additional processing step, as you suggest. As git itself is also doing it, it might not even be 'hacky' after all.

If you would like to give it a shot, I would very much welcome a PR to start hammering home a fix.
